### PR TITLE
[Core] Service run order.

### PR DESCRIPTION
**Pull request by antmat**
subj. Moved out of #161

### CODE REVIEW

**Comment by Kobolog** [comment]
This one doesn't have the signal part, is that right?

**Comment by Kobolog** [comment]
Oh, it's in a separate one, okay.

File: include/cocaine/context.hpp [hunk 1]
```diff
@@ -90,8 +90,25 @@ struct config_t {
     };
 
     typedef std::map<std::string, component_t> component_map_t;
+    typedef std::vector<std::pair<std::string, component_t>> components_t;
+
+    struct component_order_t {
```
Comment (Kobolog) [inline include/cocaine/context.hpp]: Either move it to the implementation (no point for it being in the headers).

**Comment by Kobolog** [comment]
Okay, looks good. Now explain why Locator should run last?

**Comment by antmat** [comment]
Because it uses cluster_t which can be implemented in terms of other services (for example unicorn service cocaine/cocaine-plugins#52) and of course it's better to know about all services before announcing them. Also it was default behaviour in v11, which was dropped out for some reason (https://github.com/cocaine/cocaine-core/blob/v0.11/src/context.cpp#L323). I believe it was done accidentially.

**Comment by Kobolog** [comment]
Actually, it was no accident. I think that startup ordering is a generally bad idea, since it gives false and unnecessary guarantees about service dependency chains. In fact, services are independent and there should be no assumptions about the order they appear online. If a system cannot function without strict component ordering, that system is not ready for production in the first place. I'm going to close this PR, we can discuss it later for v0.13.

**Pull request closed without merging**